



<!DOCTYPE html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      <meta http-equiv="x-ua-compatible" content="ie=edge">
      
        <meta name="description" content="Robust amplicon denoiser (RAD) contains cluster and consensus tools.">
      
      
      
        <meta name="author" content="Ben Murrell">
      
      
        <meta name="lang:clipboard.copy" content="Copy to clipboard">
      
        <meta name="lang:clipboard.copied" content="Copied to clipboard">
      
        <meta name="lang:search.language" content="en">
      
        <meta name="lang:search.pipeline.stopwords" content="True">
      
        <meta name="lang:search.pipeline.trimmer" content="True">
      
        <meta name="lang:search.result.none" content="No matching documents">
      
        <meta name="lang:search.result.one" content="1 matching document">
      
        <meta name="lang:search.result.other" content="# matching documents">
      
        <meta name="lang:search.tokenizer" content="[\s\-]+">
      
      <link rel="shortcut icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.0, mkdocs-material-3.0.3">
    
    
      
        <title>clusterpipeline - RAD.jl</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/application.451f80e5.css">
      
      
    
    
      <script src="../assets/javascripts/modernizr.1aa3b519.js"></script>
    
    
      <link href="https://fonts.gstatic.com" rel="preconnect" crossorigin>
      
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,400i,700|Roboto+Mono">
        <style>body,input{font-family:"Roboto","Helvetica Neue",Helvetica,Arial,sans-serif}code,kbd,pre{font-family:"Roboto Mono","Courier New",Courier,monospace}</style>
      
    
    <link rel="stylesheet" href="../assets/fonts/material-icons.css">
    
      <link rel="stylesheet" href="../assets/Documenter.css">
    
    
  </head>
  
    <body dir="ltr">
  
    <svg class="md-svg">
      <defs>
        
        
          <svg xmlns="http://www.w3.org/2000/svg" width="416" height="448"
    viewBox="0 0 416 448" id="__github">
  <path fill="currentColor" d="M160 304q0 10-3.125 20.5t-10.75 19-18.125
        8.5-18.125-8.5-10.75-19-3.125-20.5 3.125-20.5 10.75-19 18.125-8.5
        18.125 8.5 10.75 19 3.125 20.5zM320 304q0 10-3.125 20.5t-10.75
        19-18.125 8.5-18.125-8.5-10.75-19-3.125-20.5 3.125-20.5 10.75-19
        18.125-8.5 18.125 8.5 10.75 19 3.125 20.5zM360
        304q0-30-17.25-51t-46.75-21q-10.25 0-48.75 5.25-17.75 2.75-39.25
        2.75t-39.25-2.75q-38-5.25-48.75-5.25-29.5 0-46.75 21t-17.25 51q0 22 8
        38.375t20.25 25.75 30.5 15 35 7.375 37.25 1.75h42q20.5 0
        37.25-1.75t35-7.375 30.5-15 20.25-25.75 8-38.375zM416 260q0 51.75-15.25
        82.75-9.5 19.25-26.375 33.25t-35.25 21.5-42.5 11.875-42.875 5.5-41.75
        1.125q-19.5 0-35.5-0.75t-36.875-3.125-38.125-7.5-34.25-12.875-30.25-20.25-21.5-28.75q-15.5-30.75-15.5-82.75
        0-59.25 34-99-6.75-20.5-6.75-42.5 0-29 12.75-54.5 27 0 47.5 9.875t47.25
        30.875q36.75-8.75 77.25-8.75 37 0 70 8 26.25-20.5
        46.75-30.25t47.25-9.75q12.75 25.5 12.75 54.5 0 21.75-6.75 42 34 40 34
        99.5z" />
</svg>
        
      </defs>
    </svg>
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" data-md-component="overlay" for="__drawer"></label>
    
      <a href="../#primary-clustering-and-pipeline-functions" tabindex="1" class="md-skip">
        Skip to content
      </a>
    
    
      <header class="md-header" data-md-component="header">
  <nav class="md-header-nav md-grid">
    <div class="md-flex">
      <div class="md-flex__cell md-flex__cell--shrink">
        <a href=".." title="RAD.jl" class="md-header-nav__button md-logo">
          
            <i class="md-icon"></i>
          
        </a>
      </div>
      <div class="md-flex__cell md-flex__cell--shrink">
        <label class="md-icon md-icon--menu md-header-nav__button" for="__drawer"></label>
      </div>
      <div class="md-flex__cell md-flex__cell--stretch">
        <div class="md-flex__ellipsis md-header-nav__title" data-md-component="title">
          
            
              <span class="md-header-nav__topic">
                RAD.jl
              </span>
              <span class="md-header-nav__topic">
                clusterpipeline
              </span>
            
          
        </div>
      </div>
      <div class="md-flex__cell md-flex__cell--shrink">
        
          
            <label class="md-icon md-icon--search md-header-nav__button" for="__search"></label>
            
<div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="query" data-md-state="active">
      <label class="md-icon md-search__icon" for="__search"></label>
      <button type="reset" class="md-icon md-search__icon" data-md-component="reset" tabindex="-1">
        &#xE5CD;
      </button>
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="result">
          <div class="md-search-result__meta">
            Type to start searching
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
          
        
      </div>
      
        <div class="md-flex__cell md-flex__cell--shrink">
          <div class="md-header-nav__source">
            


  


  <a href="https://github.com/MurrellGroup/RAD.jl/" title="Go to repository" class="md-source" data-md-source="github">
    
      <div class="md-source__icon">
        <svg viewBox="0 0 24 24" width="24" height="24">
          <use xlink:href="#__github" width="24" height="24"></use>
        </svg>
      </div>
    
    <div class="md-source__repository">
      GitHub
    </div>
  </a>

          </div>
        </div>
      
    </div>
  </nav>
</header>
    
    <div class="md-container">
      
        
      
      
      <main class="md-main">
        <div class="md-main__inner md-grid" data-md-component="container">
          
            
              <div class="md-sidebar md-sidebar--primary" data-md-component="navigation">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    <nav class="md-nav md-nav--primary" data-md-level="0">
  <label class="md-nav__title md-nav__title--site" for="__drawer">
    <a href=".." title="RAD.jl" class="md-nav__button md-logo">
      
        <i class="md-icon"></i>
      
    </a>
    RAD.jl
  </label>
  
    <div class="md-nav__source">
      


  


  <a href="https://github.com/MurrellGroup/RAD.jl/" title="Go to repository" class="md-source" data-md-source="github">
    
      <div class="md-source__icon">
        <svg viewBox="0 0 24 24" width="24" height="24">
          <use xlink:href="#__github" width="24" height="24"></use>
        </svg>
      </div>
    
    <div class="md-source__repository">
      GitHub
    </div>
  </a>

    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      


  <li class="md-nav__item">
    <a href=".." title="Home" class="md-nav__link">
      Home
    </a>
  </li>

    
      
      
      

  


  <li class="md-nav__item md-nav__item--active">
    
    <input class="md-toggle md-nav__toggle" data-md-toggle="toc" type="checkbox" id="__toc">
    
      
    
    
    <a href="./" title="clusterpipeline" class="md-nav__link md-nav__link--active">
      clusterpipeline
    </a>
    
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../consensus/" title="consensus" class="md-nav__link">
      consensus
    </a>
  </li>

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              <div class="md-sidebar md-sidebar--secondary" data-md-component="toc">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    
<nav class="md-nav md-nav--secondary">
  
  
    
  
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content">
            <article class="md-content__inner md-typeset">
              
                
                  <a href="https://github.com/MurrellGroup/RAD.jl/edit/master/docs/clusterpipeline.md" title="Edit this page" class="md-icon md-content__icon">&#xE3C9;</a>
                
                
                <p><a id='Primary-Clustering-and-Pipeline-Functions-1'></a></p>
<h1 id="primary-clustering-and-pipeline-functions">Primary Clustering and Pipeline Functions</h1>
<ul>
<li><a href="./#RAD.clusterpipeline"><code>RAD.clusterpipeline</code></a></li>
<li><a href="./#RAD.denoise"><code>RAD.denoise</code></a></li>
<li><a href="./#RAD.get_ave_dists"><code>RAD.get_ave_dists</code></a></li>
<li><a href="./#RAD.kmer_split_clustering"><code>RAD.kmer_split_clustering</code></a></li>
<li><a href="./#RAD.kmer_split_consensus"><code>RAD.kmer_split_consensus</code></a></li>
</ul>
<p><a id='RAD.denoise' href='#RAD.denoise'>#</a>
<strong><code>RAD.denoise</code></strong> &mdash; <em>Function</em>.</p>
<pre><code>denoise(seqs; &lt;keyword arguments&gt; )

                rough_radius::Float64 = 0.01,
                fine_radius::Float64 = 1.0,
                reassign::Bool = true,
                clust_multiplier::Float64 = 1.0,
                min_devs::Int64 = 6,
                initial_devs::Int64 = 20,
                user_min_clust::Int64 = 5,
                k::Int64 = 6,
                distfunc = euclidean,
                center = mean,
                verbose::Int64 = 1,
                cycle_lim = 20,
                triangle::Bool = false,
                degap_param::Bool = false,
                Polish::Bool = true,
                p_value::Float64 = 0.01
</code></pre>

<h1 id="arguments">Arguments</h1>
<ul>
<li><code>seqs</code>: array containing the sequences to be operated on</li>
<li><code>rough_radius::Float64</code>: distance between SNPs used in fine clustering</li>
<li><code>fine_radius::Float64</code>: distance between SNPs used in rough clustering</li>
<li><code>reassign::Bool = true</code>: indicates whether or not to reassign the rough singletons</li>
</ul>
<p>back to consensus sequences</p>
<ul>
<li><code>clust_multiplier::Float64 = 1.0</code>: minimum non-error cluster size</li>
<li><code>min_devs::Int64 = 6</code>: number of high standard deviation kmer to fine cluster on</li>
<li><code>initial_devs::Int64 = 20</code>: initial amount of high standard deviation kmer to test</li>
<li><code>user_min_clust::Int64 = 5</code>: minimum size of a cluster, competes with clust_multiplier</li>
<li><code>k::Int64 = 6</code>: length of kmer sequences calculated</li>
<li><code>distfunc = euclidean</code>: type of distance function calculated on the kmer sequences</li>
<li><code>center = mean</code>: sets the center of a centroid to the mean of the centroid</li>
<li><code>verbose::Int64 = 1</code>: when true, the status of the program will be printed to 'stdout'</li>
<li><code>cycle_lim = 20</code>: maximum number of cycles allowed in clustering call</li>
<li><code>triangle::Bool = false</code>: controls whether or not the triangle inequality is used in calculations</li>
<li><code>degap_param::Bool = false</code>: controls whether degap is used in consensus calculations</li>
<li><code>Polish::Bool = false</code>: if true, draws a consensus again from singletons after reassignment</li>
<li><code>p_value::Float64 = 0.01</code>: statistical threshold for splitting decisions</li>
</ul>
<p>Performs clustering and sequence consensus.</p>
<p>Order of process:</p>
<ul>
<li>Rough clustering</li>
<li>Fine clustering</li>
<li>Consensus of clusters</li>
</ul>
<p>Recommended <code>radius</code> for rough clustering: 0.01 Recommended <code>radius</code> for fine clustering: 1.0</p>
<p><a target='_blank' href='https://github.com/MurrellGroup/RAD.jl/blob/6151ae06c78da4f896cc88d797fefcac436053eb/src/clusterpipeline.jl#L3-L54' class='documenter-source'>source</a><br></p>
<p><a id='RAD.kmer_split_clustering' href='#RAD.kmer_split_clustering'>#</a>
<strong><code>RAD.kmer_split_clustering</code></strong> &mdash; <em>Function</em>.</p>
<pre><code>kmer_split_clustering(kmer_vecs, seqs; rough_radius::Float64 = 0.01, fine_radius::Float64 = 1.0,  reassign::Bool = true,
                    clust_multiplier::Float64 = 1.0, min_devs::Int64 = 6, initial_devs::Int64 = 20,
                    user_min_clust::Int64 = 5, k::Int64 = 6, distfunc = euclidean, center = mean,
                    verbose::Int64 = 1, cycle_lim = 20, triangle::Bool = false, p_value::Float64 = 0.01)
</code></pre>

<p>Returns the fine clustering result from the given sequences.</p>
<p><a target='_blank' href='https://github.com/MurrellGroup/RAD.jl/blob/6151ae06c78da4f896cc88d797fefcac436053eb/src/clusterpipeline.jl#L72-L80' class='documenter-source'>source</a><br></p>
<p><a id='RAD.kmer_split_consensus' href='#RAD.kmer_split_consensus'>#</a>
<strong><code>RAD.kmer_split_consensus</code></strong> &mdash; <em>Function</em>.</p>
<pre><code>kmer_split_consensus(kmer_vecs, seqs, fine_indices; verbose::Int64=1, distfunc=euclidean, reassign::Bool=true,
                   k::Int64=6, degap_param::Bool=false, user_min_clust::Int64=5, Polish::Bool=true)
</code></pre>

<p>Returns the consensus of the sequences after clustering.</p>
<p><a target='_blank' href='https://github.com/MurrellGroup/RAD.jl/blob/6151ae06c78da4f896cc88d797fefcac436053eb/src/clusterpipeline.jl#L93-L99' class='documenter-source'>source</a><br></p>
<p><a id='RAD.get_ave_dists' href='#RAD.get_ave_dists'>#</a>
<strong><code>RAD.get_ave_dists</code></strong> &mdash; <em>Function</em>.</p>
<pre><code>get_ave_dists(fine_indices, original_seqs, consensus_seqs; distfunc=euclidean, center=median, k::Int64=6)
</code></pre>

<p>Finds the average distance using the declared distance function</p>
<p><a target='_blank' href='https://github.com/MurrellGroup/RAD.jl/blob/6151ae06c78da4f896cc88d797fefcac436053eb/src/clusterpipeline.jl#L148-L152' class='documenter-source'>source</a><br></p>
<p><a id='RAD.clusterpipeline' href='#RAD.clusterpipeline'>#</a>
<strong><code>RAD.clusterpipeline</code></strong> &mdash; <em>Function</em>.</p>
<pre><code>clusterpipeline(infile::String; &lt;Keyword Arguments&gt; )

                     prefix = &quot;&quot;,
                     outfile = &quot;&quot;,
                     ref_file = &quot;&quot;,
                     error_filter = 0.01,
                     homopoly_filter = true,
                     minlength = 20,
                     maxlength = 1000000,
                     trim_seqs = true,
                     align_ref_frames = true,
                     write_bad_seqs = false,
                     write_aminoacids = true,
                     minsize = 5,
                     cycle_lim = 10,
                     merge_differing_homopolymers = false,
                     cluster_filedump_path = &quot;&quot;,
                     post_filedump = false,
                     rough_radius::Float64=0.01,
                     fine_radius::Float64=1.0,
                     reassign::Bool=true,
                     clust_multiplier::Float64=1.0,
                     min_devs::Int64=6,
                     initial_devs::Int64=20,
                     user_min_clust::Int64=5,
                     k::Int64=6,
                     distfunc=euclidean,
                     center=mean,
                     verbose::Int64=1,
                     triangle::Bool=false,
                     degap_param::Bool=false,
                     Polish::Bool=true,
                     p_value::Float64=0.01)
</code></pre>

<p><strong>Arguments</strong></p>
<ul>
<li><code>prefix = ""</code>: prefix for names of sequences in output file (ie in case of multiple time points, etc.).</li>
<li><code>outfile = ""</code>: default is "[infile name without extension]-clusters.fasta".</li>
<li><code>ref_file = ""</code>: panel of reference seqs – if provided, used to orient and trim nearest sequences in input file.</li>
<li><code>error_filter = 0.01</code>: if the input is a ".fastq" file, any sequence with a site with error greater than this value is removed before clustering.</li>
<li><code>homopoly_filter = true</code>: if true, remove sequences with erroneous run-on homopolymer regions (or trim the sequences if this region is on an end) via hidden markov model inference.</li>
<li><code>minlength = 20</code>: removes sequences shorter than this prior to clustering.</li>
<li><code>maxlength = 1000000</code>: removes sequences longer than this prior to clustering.</li>
<li><code>trim_seqs = true</code>: if a reference panel is provided (<code>ref_file</code>) and this is true, locally trims consensus seqs to nearest reference sequence</li>
<li><code>align_ref_frames = true</code>: corrects codon reading frames of inferred consensus sequences.</li>
<li><code>write_bad_seqs = false</code>: write sequences with uncorrectable reading frames to file</li>
<li><code>write_aminoacids = true</code>: write amino acid translations of consensus sequences to file</li>
<li><code>minsize = 5</code>: throw away clusters/consensuses with fewer than this many reads.</li>
<li><code>cycle_lim = 10</code> : TODO</li>
<li><code>merge_differing_homopolymers = false</code>: aggregates clusters with consensus seqs that only differ by a single gap in a homopolymer region.</li>
<li><code>cluster_filedump_path = ""</code>: if nonempty, writes original sequences to output files determined by resulting cluster in given directory. This string may also include a file name prefix.</li>
<li><code>post_filedump</code>: writes inferred consensus seqs to file after clustering, before aligning and trimming.</li>
<li><code>rough_radius::Float64</code>: distance between SNPs used in fine clustering</li>
<li><code>fine_radius::Float64</code>: distance between SNPs used in rough clustering</li>
<li><code>reassign::Bool = true</code>: indicates whether or not to reassign the rough singletons</li>
</ul>
<p>back to consensus sequences</p>
<ul>
<li><code>clust_multiplier::Float64 = 1.0</code>: minimum non-error cluster size</li>
<li><code>min_devs::Int64 = 6</code>: number of high standard deviation kmer to fine cluster on</li>
<li><code>initial_devs::Int64 = 20</code>: initial amount of high standard deviation kmer to test</li>
<li><code>user_min_clust::Int64 = 5</code>: minimum size of a cluster, competes with clust_multiplier</li>
<li><code>k::Int64 = 6</code>: length of kmer sequences calculated</li>
<li><code>distfunc = euclidean</code>: type of distance function calculated on the kmer sequences</li>
<li><code>center = mean</code>: sets the center of a centroid to the mean of the centroid</li>
<li><code>verbose::Int64 = 1</code>: when true, the status of the program will be printed to 'stdout'</li>
<li><code>cycle_lim = 20</code>: maximum number of cycles allowed in clustering call</li>
<li><code>triangle::Bool = false</code>: controls whether or not the triangle inequality is used in calculations</li>
<li><code>degap_param::Bool = false</code>: controls whether degap is used in consensus calculations</li>
<li><code>Polish::Bool = false</code>: if true, draws a consensus again from singletons after reassignment</li>
<li><code>p_value::Float64 = 0.01</code>: statistical threshold for splitting decisions</li>
</ul>
<p>A clustering pipeline for sequence files. Order of processes:</p>
<ul>
<li>Error filter (if .fastq)</li>
<li>Length filter</li>
<li>Homopolymer filter</li>
<li>Length filter again</li>
<li>Orient (either to reference panel or inferred coarse cluster centroid)</li>
<li>Rough clustering</li>
<li>Fine clustering</li>
<li>Consensus of clusters</li>
</ul>
<p>Recommended <code>radius</code> for rough clustering: 0.01 Recommended <code>radius</code> for fine clustering: 1.0</p>
<p><code>infile</code> and <code>ref_file</code> must be either .fasta or .fastq. If file is .fastq, then removes sequences with greater than <code>error_filter</code> error.</p>
<p><a target='_blank' href='https://github.com/MurrellGroup/RAD.jl/blob/6151ae06c78da4f896cc88d797fefcac436053eb/src/clusterpipeline.jl#L166-L257' class='documenter-source'>source</a><br></p>
                
                  
                
              
              
                


              
            </article>
          </div>
        </div>
      </main>
      
        
<footer class="md-footer">
  
    <div class="md-footer-nav">
      <nav class="md-footer-nav__inner md-grid">
        
          <a href=".." title="Home" class="md-flex md-footer-nav__link md-footer-nav__link--prev" rel="prev">
            <div class="md-flex__cell md-flex__cell--shrink">
              <i class="md-icon md-icon--arrow-back md-footer-nav__button"></i>
            </div>
            <div class="md-flex__cell md-flex__cell--stretch md-footer-nav__title">
              <span class="md-flex__ellipsis">
                <span class="md-footer-nav__direction">
                  Previous
                </span>
                Home
              </span>
            </div>
          </a>
        
        
          <a href="../consensus/" title="consensus" class="md-flex md-footer-nav__link md-footer-nav__link--next" rel="next">
            <div class="md-flex__cell md-flex__cell--stretch md-footer-nav__title">
              <span class="md-flex__ellipsis">
                <span class="md-footer-nav__direction">
                  Next
                </span>
                consensus
              </span>
            </div>
            <div class="md-flex__cell md-flex__cell--shrink">
              <i class="md-icon md-icon--arrow-forward md-footer-nav__button"></i>
            </div>
          </a>
        
      </nav>
    </div>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">
        
        powered by
        <a href="https://www.mkdocs.org">MkDocs</a>
        and
        <a href="https://squidfunk.github.io/mkdocs-material/">
          Material for MkDocs</a>
      </div>
      
        
      
    </div>
  </div>
</footer>
      
    </div>
    
      <script src="../assets/javascripts/application.583bbe55.js"></script>
      
      <script>app.initialize({version:"1.0",url:{base:".."}})</script>
      
        <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS_HTML"></script>
      
        <script src="../assets/mathjaxhelper.js"></script>
      
    
    
      
    
  </body>
</html>